### DISCUSSION TITLE
Is it possible to allow controllers to be internal inside of modules?

### DISCUSSION THREAD

**Discussion by ChristopherHaws** [OrchardCMS/OrchardCore#11417] [Q&A]
I am trying to get Orchard to expose internal controllers within my modules but can't seem to figure out how. In standard ASP.NET Core, the way I do it is by replacing the `ControllerFeatureFactory` with a custom one:

```csharp
builder.Services.AddMvc().ConfigureApplicationPartManager(mvc => {
   // Add feature provider to allow "internal" controller
   mvc.FeatureProviders.Add(new InternalControllerFeatureProvider());
});

internal class InternalControllerFeatureProvider : ControllerFeatureProvider {
    protected override bool IsController(TypeInfo typeInfo) {
        var isCustomController = !typeInfo.IsAbstract
            && typeof(ControllerBase).IsAssignableFrom(typeInfo)
            && IsInternal(typeInfo);
        return isCustomController || base.IsController(typeInfo);

        bool IsInternal(Type t) =>
            !t.IsVisible
            && !t.IsPublic
            && t.IsNotPublic
            && !t.IsNested
            && !t.IsNestedPublic
            && !t.IsNestedFamily
            && !t.IsNestedPrivate
            && !t.IsNestedAssembly
            && !t.IsNestedFamORAssem
            && !t.IsNestedFamANDAssem;
    }
}
```

Is there some way to register this in Orchard Core or another way to achieve the same result?

Thanks,
Chris

**Comment by jtkech** [comment]
In OrchardCore most of the things run in a tenant context, you can have multi tenants composed of enabled modules, and at least we have a Default tenant, each tenant has an isolated part manager so here normally you can register what you want but at the tenant level, can be done by the startup of a dedicated module or from the app by using one of our helpers.

So not fully sure but try to do what you are used to do from the application startup, but by using `ConfigureServices()` helper acting on an `OrchardCoreBuilder`, this to register your feature provider at the tenant level.

        public void ConfigureServices(IServiceCollection services)
        {
            // Or AddOrchardCore() if not using the cms.
            services.AddOrchardCms()
                // Helper acting on an OrchardCoreBuilder returned by AddOrchardCms() or AddOrchardCore().
                .ConfigureServices(tenantServices =>
                {
                    // Addmvc() is called for you at the tenant level, but you can re-call it.
                    tenantServices.AddMvc()
                    .ConfigureApplicationPartManager(partManager =>
                    {
                        partManager.FeatureProviders.Add(...);
                    });
                })
                .AddSetupFeatures("OrchardCore.AutoSetup");
        }

**Reply by ChristopherHaws** [reply]
@jtkech Yes, this is exactly what I tried (along with trying to configure it in the module itself) but unfortunately it doesn't seem to work. Public controllers resolve fine, but internal one still cannot be accessed.

Here is where orchard configures the part manager, but I don't see any extensibility points for ControllerFeature's (I do for ViewFeature's however): https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Mvc.Core/Startup.cs#L138

For reference, here is what my host's Program.cs looks like:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOrchardCore().AddMvc().ConfigureServices(tenantServices => {
    tenantServices.AddMvc().ConfigureApplicationPartManager(partManager => {
        partManager.FeatureProviders.Add(new InternalControllerFeatureProvider());
    });
});

var app = builder.Build();

app.UseHttpsRedirection();
app.UseStaticFiles();

app.UseOrchardCore();

app.Run();

internal class InternalControllerFeatureProvider : ControllerFeatureProvider {
    protected override bool IsController(TypeInfo typeInfo) {
        var isCustomController = !typeInfo.IsAbstract
            && typeof(ControllerBase).IsAssignableFrom(typeInfo)
            && IsInternal(typeInfo);
        return isCustomController || base.IsController(typeInfo);

        bool IsInternal(Type t) =>
            !t.IsVisible
            && !t.IsPublic
            && t.IsNotPublic
            && !t.IsNested
            && !t.IsNestedPublic
            && !t.IsNestedFamily
            && !t.IsNestedPrivate
            && !t.IsNestedAssembly
            && !t.IsNestedFamORAssem
            && !t.IsNestedFamANDAssem;
    }
}
```

**Reply by ChristopherHaws** [reply]
After debugging through the code a little bit, it looks like [this `.Where` statement](https://github.com/OrchardCMS/OrchardCore/blob/743ecc6a0d186803e8b152e5585acee1a7442223/src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs#L285) is causing the internal controller to be filtered out because [`IsComponentType` filters out non-public types](https://github.com/OrchardCMS/OrchardCore/blob/743ecc6a0d186803e8b152e5585acee1a7442223/src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs#L343). Is there a reason for this? :)

**Comment by Piedone** [comment]
While I won't be able to help more than JT, can you explain what you'd like to achieve by keeping the controllers internal but still getting them directly registered?

**Reply by ChristopherHaws** [reply]
@Piedone We do not want to expose types from our modules that might accidently be referenced by other modules/libraries. For example, one module might have an EF Core DbContext that has all of it's entities defined in it. We don't want that to be public for anyone to use, but if the controller is public and we are using IoC, then all of its dependencies also need to be made public. Unfortunately, `Microsoft.Extensions.DependencyInjection` doesn't allow you to make constructors internal, so instead we need to make the class internal and keep the constructor public. I am not sure if this is the case with other IoC containers, but MS did give some good reasons why they made this decision in a github issue (I could dig it up if you want).

Here is an article that talks about modular monolith design where they talk about using internal types over public types:
https://www.thinktecture.com/en/asp-net/modular-monolith/#customizing-aspnet-core

Maybe this could be made into a configurable option when setting up orchard (i.e. `.AddOrchardCore(x => x.AllowInternalTypes = true)`?

**Reply by Piedone** [reply]
Interesting, thank you for the explanation. And yes, this could very well be a feature, please open an issue here: https://github.com/OrchardCMS/OrchardCore/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=

**Reply by ChristopherHaws** [reply]
Here is a link to the issue I made since it doesn't look like GitHub adds a link automatically like it does in issues. https://github.com/OrchardCMS/OrchardCore/issues/11426

**Reply by Piedone** [reply]
Thanks!
